tests/ClientTest.php

```diff
@@ -0,0 +1,292 @@
+            ),
+        ]);
+        // Add the history middleware to the handler stack.
+        $requestHistory = [];
+        $history = Middleware::history($requestHistory);
+        $stack = HandlerStack::create($mock);
+        $stack->push($history);
+        $client = $this->getClient(['handler' => $stack]);
+        $client->postEvent($this->getPostEventData());
+        self::assertCount(3, $requestHistory);
+        /** @var Request $request */
+        $request = $requestHistory[0]['request'];
+        self::assertEquals('http://example.com/events/job_failed', $request->getUri()->__toString());
+        $request = $requestHistory[1]['request'];
+        self::assertEquals('http://example.com/events/job_failed', $request->getUri()->__toString());
+        $request = $requestHistory[2]['request'];
+        self::assertEquals('http://example.com/events/job_failed', $request->getUri()->__toString());
+    }
+
+    public function testRetryFailure(): void
+    {
+        $responses = [];
+        for ($i = 0; $i < 30; $i++) {
+            $responses[] = new Response(
+                500,
+                ['Content-Type' => 'application/json'],
+                '{"message" => "Out of order"}'
+            );
+        }
+        $mock = new MockHandler($responses);
+        // Add the history middleware to the handler stack.
+        $requestHistory = [];
+        $history = Middleware::history($requestHistory);
+        $stack = HandlerStack::create($mock);
+        $stack->push($history);
+        $client = $this->getClient(['handler' => $stack, 'backoffMaxTries' => 1]);
+        try {
+            $client->postEvent($this->getPostEventData());
+            self::fail('Must throw exception');
+        } catch (ClientException $e) {
+            self::assertStringContainsString('500 Internal Server Error', $e->getMessage());
+        }
+        self::assertCount(2, $requestHistory);
+    }
+
+    public function testRetryFailureReducedBackoff(): void
+    {
+        $responses = [];
+        for ($i = 0; $i < 30; $i++) {
+            $responses[] = new Response(
+                500,
+                ['Content-Type' => 'application/json'],
+                '{"message" => "Out of order"}'
+            );
+        }
+        $mock = new MockHandler($responses);
+        // Add the history middleware to the handler stack.
+        $requestHistory = [];
+        $history = Middleware::history($requestHistory);
+        $stack = HandlerStack::create($mock);
+        $stack->push($history);
+        $client = $this->getClient(['handler' => $stack, 'backoffMaxTries' => 3]);
+        try {
+            $client->postEvent($this->getPostEventData());
+            self::fail('Must throw exception');
+        } catch (ClientException $e) {
+            self::assertStringContainsString('500 Internal Server Error', $e->getMessage());
+        }
+        self::assertCount(4, $requestHistory);
+    }
+
+    public function testNoRetry(): void
```

Comment:
"no retry" je to diky tomu, ze response je `401` nebo kvuli tomu, ze nemam nakonfigurovany `backoffMaxTries`?

Reply:
kvůli tomu, že to je 400, retries jsou by default  - viz testRetrySuccess